In [4]:
import wandb
wandb.login()

wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice:

  2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.
wandb: Paste your API key and hit enter:

  ········


wandb: ERROR Invalid API key: API key must have 40+ characters, has 36.
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice:

  2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.
wandb: Paste your API key and hit enter:

  ········


wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: C:\Users\THIEN\_netrc
wandb: Currently logged in as: 2351010189tam (2351010189tam-notion) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

In [6]:
import numpy as np
import matplotlib.pyplot as plt

from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

from preprocessing import load_and_split, build_preprocessor


X_train, X_test, y_train, y_test = load_and_split("../data/student-mat.csv")


preprocessor = build_preprocessor()


configs = [
    {"n_estimators": 100, "max_depth": None},
    {"n_estimators": 100, "max_depth": 5},
    {"n_estimators": 200, "max_depth": None},
    {"n_estimators": 200, "max_depth": 5},
    {"n_estimators": 300, "max_depth": 10},
]

for cfg in configs:

    wandb.init(
        project="student-performance",
        config={
        "model": "RandomForestRegressor",
        "n_estimators": cfg["n_estimators"],
        "max_depth": cfg["max_depth"],
        "random_state": 42
    }

    )


    pipeline = Pipeline([
        ("preprocessing", preprocessor),
        ("model", RandomForestRegressor(
            n_estimators=cfg["n_estimators"],
            max_depth=cfg["max_depth"],
            random_state=42
        ))
    ])


    pipeline.fit(X_train, y_train)


    y_pred = pipeline.predict(X_test)


    mae = mean_absolute_error(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)


    wandb.log({
        "MAE": mae,
        "RMSE": rmse,
        "R2": r2
    })


    plt.figure(figsize=(6, 6))
    plt.scatter(y_test, y_pred)
    plt.xlabel("Actual")
    plt.ylabel("Predicted")
    plt.title("Actual vs Predicted")

    wandb.log({
        "Prediction Chart": wandb.Image(plt)
    })

    plt.close()


    wandb.finish()

MAE,▁
R2,▁
RMSE,▁
MAE,1.4513
R2,0.82233
RMSE,1.90869


MAE,▁
R2,▁
RMSE,▁
MAE,1.53143
R2,0.80141
RMSE,2.01792


MAE,▁
R2,▁
RMSE,▁
MAE,1.4367
R2,0.8243
RMSE,1.89811


MAE,▁
R2,▁
RMSE,▁
MAE,1.52923
R2,0.79967
RMSE,2.02677


MAE,▁
R2,▁
RMSE,▁
MAE,1.44346
R2,0.82465
RMSE,1.89621


Đã thực hiện 5 thí nghiệm để so sánh hiệu suất các mô hình.

Experiment 1: RMSE = 1.90869, MAE = 1.4513, R² = 0.82233. Kết quả khá tốt.
Experiment 2: RMSE = 2.01792, MAE = 1.53143, R² = 0.80141. Hiệu suất giảm nhẹ.
Experiment 3: RMSE = 1.89811, MAE = 1.4367, R² = 0.8243. Kết quả tốt nhất.
Experiment 4: RMSE = 2.02677, MAE = 1.52923, R² = 0.79967. Kết quả thấp nhất.
Experiment 5: RMSE = 1.89621, MAE = 1.44346, R² = 0.82465. Kết quả rất tốt và ổn định.
Kết luận

Experiment 5 có R² cao nhất và RMSE thấp nhất, cho thấy khả năng dự đoán tốt nhất.
Nhìn chung, các mô hình có hiệu suất khá ổn định với R² khoảng 0.80–0.82.